org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -77,24 +80,44 @@ public CodeActionHandler(PreferenceManager preferenceManager) {
 		IProblemLocationCore[] locations = this.getProblemLocationCores(unit, params.getContext().getDiagnostics());
 
 		List<Either<Command, CodeAction>> $ = new ArrayList<>();
+		List<CUCorrectionProposal> candidates = new ArrayList<>();
 		try {
-			CUCorrectionProposal[] corrections = this.quickFixProcessor.getCorrections(context, locations);
-			Arrays.sort(corrections, new CUCorrectionProposalComparator());
-			for (CUCorrectionProposal proposal : corrections) {
-				$.add(getCodeActionFromProposal(proposal, params.getContext()));
-			}
+			ArrayList<CUCorrectionProposal> corrections = this.quickFixProcessor.getCorrections(context, locations);
+			corrections.sort(new CUCorrectionProposalComparator());
+			candidates.addAll(corrections);
+		} catch (CoreException e) {
+			JavaLanguageServerPlugin.logException("Problem resolving quick fix code actions", e);
+		}
+
+		try {
+			ArrayList<CUCorrectionProposal> corrections = this.quickAssistProcessor.getAssists(context, locations);
+			corrections.sort(new CUCorrectionProposalComparator());
+			candidates.addAll(corrections);
 		} catch (CoreException e) {
-			JavaLanguageServerPlugin.logException("Problem resolving code actions", e);
+			JavaLanguageServerPlugin.logException("Problem resolving quick assist code actions", e);
+		}
+
+		ArrayList<CUCorrectionProposal> corrections = this.sourceAssistProcessor.getAssists(context, locations);
+		corrections.sort(new CUCorrectionProposalComparator());
```

Comment:
sort candidates instead

Reply:
Would keep the group by kind before sort.

Reply:
then just make the comparator use the kind 1st